### ISSUE TITLE
Doesn't work if closed from the "Quit" menu entry in Firefox

### ISSUE THREAD

**Issue by MilaMurry** [M-Reimer/cookiewhitelist#1]
It seems it doesn't work as it intended to be.

**Scenario 1**
Go to [your AMO](https://addons.mozilla.org/en-US/firefox/addon/cookiewhitelist/), then close the tab, visit this [link (farside)](https://farside.link/), open the add-on, and you'll see there's a Mozilla cookie/localStorage in the list.

**Scenario 2**
Follow the scenario 1, close the browser, visit [farside](https://farside.link/), open the add-on and you'll see the same thing.

I noticed it will only be triggered when I force stop the browser, reopen it and visit whatever website.

Well, I'm not sure when will it run either when a tab or browser closes, but I'd like to have both options in the settings. But don't delete just yet when multiple tabs still using the same domain.

Regardless, it would be better to have separate tabs for each category as `cookies`, `localStorage`, `indexedDB` and `all`.

Also please add these 2 words: `localStorage` and `indexedDB` in your add-on's description to make it clear. So the description may sound like this:

*Deletes all cookies, localStorage and indexedDB from domains, not added to the provided whitelist, as soon as you exit your browser.*

Last but not least, I'm using [Iceraven](https://github.com/fork-maintainers/iceraven-browser) browser if that makes it different.

**Comment by M-Reimer** [comment]
You have to close the browser to clear cookies. This is intended behavior. There are other Add-ons which work "by tab" but this approach was not very reliable in my tests.

You don't have to "force close". Just close the browser in the list of recently used apps (slide from the lower edge upwards).

Separate tabs would not work. The whitelist does not apply to localStorage and indexedDB.

As your description is wrong and just to make it clear: The whitelist is only applied to cookies. If you enable the deletion localStorage and indexedDB, then everything is deleted there! There is no other way to do it. Yes, I have to document this better and I will link some Mozilla bugs which are the reason for this limitations.

**Comment by MilaMurry** [comment]
So the deletion of the localStorage and indexedDB isn't based on the whitelist? I thought so. I know that you can clear the localStorage by tab using `localStorage.clear()`, so it might work if it has that functionality.

And before I forget, other add-ons like `CAD` or `Forget Me Not` don't work properly, which is why I'm testing this add-on.

Unfortunately, closing and removing the app from recently used apps doesn't work either.

Well whatever the case, you can safely close this issue.

**Comment by M-Reimer** [comment]
Mozilla does not have API to apply a whitelist when cleaning localStorage or indexedDB. But mobile Firefox doesn't have any other way to clear them built in. So I added the checkbox. I would have printed the "all domains" in bold but my translation library does not support that currently.

Clearing localStorage by tab only works if the tab is still open and it is not reliable and impossible to check. Mozilla does not have API to tell if localStorage is in use or which domain uses it. So to be sure I decided clearing all could be a possible solution.

This, for sure, has disadvantages. I already have a report that you get logged out of Google accounts as they seem to store the session in localStorage. If this gets more common, this would render my Add-on useless. But at least for now I keep logged in where I want to.

I'll keep this issue open as a reminder that I still have to improve documentation including some links to the relevant bugs.

The big problem with Firefox Add-ons is that Add-on API seems to be at the very end of the Mozilla priority list. It is pretty common for bugs being open for years. So it's all about getting something done with what you have and find suitable workarounds for bugs in their API as you can never count on anything getting fixed in time.

**Comment by BD9Max** [comment]
Works great in Firefox Android 👍👍👏

**Comment by MilaMurry** [comment]
I don't think it actually deletes the real files. Recently Android reminded me that the browser was using too much storage and indeed it was. Luckily my device was rooted. I had to delete a bunch of folders  (localStorage) like `https+++github.com`, `https+++youtube.com`, etc...

Of course this add-on works just fine, but like you said it's on Firefox's side. JFYI.

**Comment by jra66** [comment]
Well, the add-on cleans the cookies, but it is triggered by closing the browser via the list of recently opened apps only. When I close the browser with the menu entry the cookies won't be deleted. I prefer to close Firefox using the menu because I have configured it to erase history, downloads etc. when closed. I unchecked cookies in this list in order to be deleted by the add-on, respecting the white-list which I need to keep the cookies for the startpage.com search engine.

Is it possible that the add-on gets triggered by closing the browser via the menu, too? That would be great!

Kind regards,

Jörg

**Comment by M-Reimer** [comment]
What is this "menu". I don't have any "menu" to close an app in.

**Comment by M-Reimer** [comment]
Found it. The menu entry is not there by default.

And I checked documentation and as far as I could find out it is not possible to hook into this "Close" menu entry.

So: No, not possible. I also don't see a big benefit in deleting those as none of the additional items, that could be cleared by Firefox, allow tracking.

**Comment by M-Reimer** [comment]
Actually the cookies should be deleted.
Closing Firefox using the menu item actually removes it from the list of active apps.
And my Add-on clears cookies on the next Firefox startup.
Are you sure this doesn't work for you?

**Comment by jra66** [comment]
Hi Manuel,

Thanks a lot for your work. In fact, the add-on doesn't work for me as expected. My settings are as follows:

![Image](https://github.com/user-attachments/assets/a16ae84d-cfa4-4327-97db-991cb12eeaad)

When I start Firefox and go to Amazon your add-on shows these cookies:

![Image](https://github.com/user-attachments/assets/dd129377-5465-4886-b0fa-b9995c7a4161)

After closing Firefox via the menu and restarting it, no tabs are open anymore:

![Image](https://github.com/user-attachments/assets/db6ea314-e06f-4515-bdd3-f11004d19f5c)

Then I go to eBay:

![Image](https://github.com/user-attachments/assets/f5f051e5-73bf-45f6-8c30-4081fcc65a31)

You can see that Amazon's cookies are still there in addition to eBay's ones. Maybe it is something wrong with Firefox itself because cookies of Wikipedia (from the past?) show up though I didn't open Wikipedia in this session.

Kind regards,

Jörg

**Comment by M-Reimer** [comment]
Just checked with a debugger and Firefox DOES NOT actually exit if you choose "Exit" from that menu. I'll check which other options I might have.

**Comment by M-Reimer** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1961638
Giving up. Upstream bug. Workaround: Close Firefox from the list of recently opened apps.

**Comment by jra66** [comment]
Hi Manuel,

Thanks a lot for your work and opening the issue at Mozilla.

Kind regards,

Jörg

**Comment by guidot** [comment]
Hi Jörg,

I'm piggy packing here because the title fits:

I'm using Fennec 143.0.3 (on Android 15) and for the addon to delete the cookies I sometimes have to close the app _two_ times.

I was also curious to know if the cookies get cleared if the app is closed by the OS.

**Comment by M-Reimer** [comment]
You can see if cookies have been cleared if you open the configuration page of the Add-on directly after restarting.

For clarification: Clearing does not actually happen if Firefox is closed. This is technically impossible as Firefox does not allow to block the "app shutdown process". So cookie deletion would not be guaranteed to actually complete before the clearing process gets terminated. To work around this limitation, cookies are actually cleared at the next startup. This, unfortunately, is also not perfect which is why I have opened a bug at Mozilla.

Creating "cool Add-ons" for Firefox usually is (and actually always has been) a creative process of working around all kinds of issues. While filing bugs actually is the correct process to let Mozilla know about their issues, in my experience most bugs, only targeting the Add-on API, are more or less ignored by Mozilla.
